### PR TITLE
Combine withEffect PathEffect definitions.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

could be used in #15458 as well.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by anntzer** [comment]
Re docs: good catch, this is actually a limitation of the implementation of inspect.getdoc (which can be fixed at least in this case), I'll open a bug report on the python tracker.  In the meantime I just inherited the docstring manually.
Re naming: `_make_with_effect_patheffectsubclass`?  Happy for more proposals.

**Comment by timhoffm** [comment]
What about `withStroke = _subclass_with_normal(effect_class=Stroke)`?

**Comment by anntzer** [comment]
Sounds good, changed accordingly.

**Comment by anntzer** [comment]
(the docstring inheritance issue is at https://bugs.python.org/issue38603)

File: lib/matplotlib/patheffects.py [hunk 1]
```diff
@@ -159,7 +159,26 @@ class Normal(AbstractPathEffect):
     The Normal PathEffect's sole purpose is to draw the original artist with
     no special path effect.
     """
-    pass
+
+
+def _subclass_with_normal(effect_class):
+    """
+    Create a PathEffect class combining the effect of *cls* and a normal draw.
```
Comment (timhoffm) [inline lib/matplotlib/patheffects.py]:
```suggestion
    Create a PathEffect class combining the effect of *effect_class* and a normal draw.
```

File: lib/matplotlib/patheffects.py [hunk 2]
```diff
@@ -159,7 +159,26 @@ class Normal(AbstractPathEffect):
...
+    """
+
+    class withEffect(effect_class):
+        def draw_path(self, renderer, gc, tpath, affine, rgbFace):
+            super().draw_path(renderer, gc, tpath, affine, rgbFace)
+            renderer.draw_path(gc, tpath, affine, rgbFace)
+
+    withEffect.__name__ = f"with{effect_class.__name__}"
+    withEffect.__doc__ = f"""
+    Adds a `.{effect_class.__name__}` and then draws the original Artist to
```
Comment (timhoffm) [inline lib/matplotlib/patheffects.py]: I've been struggling to understand what the `with...` classes are and what they are good for (path effect newbie here). Therefore I'm a bit picky on docs. I propose something like this (Please check ReST syntax and line length. This is a bit difficult in the github interface):

~~~
"""
A shortcut PathEffect for applying `.{effect_class.__name__}` and then
drawing the original Artist.

Examples
--------
With this class you can use ::

    artist.set_path_effects([path_effects.with{effect_class.__name__}()])

as a shortcut for ::

    artist.set_path_effects([path_effects.{effect_class.__name__}(),
                             path_effects.Normal()])
"""
~~~

**Comment by timhoffm** [comment]
Bonus points if you add a link to https://matplotlib.org/3.1.1/tutorials/advanced/patheffects_guide.html in the module docstring.

File: lib/matplotlib/patheffects.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/patheffects.py]: Taking for example the path effect proposed at #15458, using just `path_effects=[TickedStroke(...)]` would *only* draw the ticks, not the "original" path; using `path_effects=[withTickedStroke(...)]` draws both the original path and not the Stroke.  Your suggestion is fine, adding it.

**Comment by anntzer** [comment]
all suggestions integrated.

File: lib/matplotlib/patheffects.py [hunk 3]
```diff
@@ -159,7 +161,35 @@ class Normal(AbstractPathEffect):
...
+    """
+    Create a PathEffect class combining *effect_class* and a normal draw.
+    """
+
+    class withEffect(effect_class):
+        def draw_path(self, renderer, gc, tpath, affine, rgbFace):
+            super().draw_path(renderer, gc, tpath, affine, rgbFace)
+            renderer.draw_path(gc, tpath, affine, rgbFace)
+
+    withEffect.__name__ = f"with{effect_class.__name__}"
```
Comment (jklymak) [inline lib/matplotlib/patheffects.py]: So... if you have to do this big song and dance for the docs, is this change really worth it?

Comment (anntzer) [reply lib/matplotlib/patheffects.py]: It's more worth it if we keep adding new patheffects and repeat this code pattern again and again, e.g. https://github.com/matplotlib/matplotlib/pull/15458.

**Comment by anntzer** [comment]
rebased

**Pull request merged**